### ISSUE TITLE
Notification syncing hampered by thread mutes

### ISSUE THREAD

**Issue by dianna-io** [bluesky-social/atproto#952]
You guessed it, hellthread strikes once again.

I've been wondering for some time why notifications on the iOS and web clients (presumably others as well) go all wonky / silent when I mute hellthread, but then come back just fine when I unmute hellthread. I realized what must be happening after some judicious spamming of refresh on notifications in either client would eventually backfill the (visible, unmuted) notification list.

I went spelunking through https://github.com/bluesky-social/indigo/tree/main/api/bsky and checked out https://github.com/bluesky-social/indigo/blob/main/api/bsky/notificationlistNotifications.go, and noticed that mutes are not applied server side, and without the client source available I can only guess right now that the filtering is happening client-side. When hellthread comes into play, I suppose this also explains why the client performance degrades with these huge mute lists.

This is something of a nasty problem to handle, isn't it? Off the top of my head I can't think of good solutions that don't just naively move the inefficiencies somewhere else.

**Comment by dholms** [comment]
Yeah I'm planning to move thread mutes to the server but have not had the bandwidth to get it done:
https://github.com/bluesky-social/atproto/pull/841

Soon :)

**Comment by dianna-io** [comment]
In the light of day and with full wakefulness I believe this could be adequately addressed with better client-side caching of notifications and the cursor (it seems to get wiped pretty easily?), along with some judicious background sync'ing when the client observes that it's not actually populating any notifications for the user after the filtered list ends up being empty, or near empty, and the clients knows there remains more notifications to fetch (which I think(?) is possible to know through `getUnreadCount`?).

**Comment by dianna-io** [comment]
I noticed the iOS client at least seems to keep fetching notifications until it has roughly a page to display after filtering, which is fantastic! thank you!
